Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/exposurewindows/AnalyticsExposureWindowDonor.kt

```diff
@@ -0,0 +1,99 @@
+package de.rki.coronawarnapp.datadonation.analytics.modules.exposurewindows
+
+import androidx.annotation.VisibleForTesting
+import de.rki.coronawarnapp.datadonation.analytics.modules.DonorModule
+import de.rki.coronawarnapp.server.protocols.internal.ppdd.PpaData
+import timber.log.Timber
+import javax.inject.Inject
+import javax.inject.Singleton
+import kotlin.random.Random
+
+@Singleton
+class AnalyticsExposureWindowDonor @Inject constructor(
+    private val analyticsExposureWindowRepository: AnalyticsExposureWindowRepository
+) : DonorModule {
+
+    override suspend fun beginDonation(request: DonorModule.Request): DonorModule.Contribution {
+        // clean up
+        analyticsExposureWindowRepository.deleteStaleData()
+
+        val probability = request.currentConfig.analytics.probabilityToSubmitNewExposureWindows
+        if (skipSubmission(probability)) {
+            Timber.w("Submission skipped.")
+            return emptyContribution
+        }
+
+        val newWrappers = analyticsExposureWindowRepository.getAllNew()
+        val reported = analyticsExposureWindowRepository.moveToReported(newWrappers)
+        return Contribution(
+            data = newWrappers.asPpaData(),
+            onDonationFailed = { onDonationFailed(newWrappers, reported) }
+        )
+    }
+
+    override suspend fun deleteData() {
+        analyticsExposureWindowRepository.deleteAllData()
+    }
+
+    @VisibleForTesting
+    internal fun skipSubmission(probability: Double): Boolean {
+        // load balancing
+        val random = Random.nextDouble()
+        Timber.w("Random number is $random. probabilityToSubmitNewExposureWindows is $probability.")
+        return random > probability
+    }
+
+    @VisibleForTesting
+    internal val emptyContribution by lazy {
+        Contribution(
+            data = emptyList(),
+            onDonationFailed = {}
+        )
```

Comment:
Q: why is wrapped in lazy ?  I don't think it a heavy resource.
and I think you should improve it, in the logs it looks like this:
```2021-02-24 13:45:09.731 13186-13364/de.rki.coronawarnapp.test I/AnalyticsExposureWindowDonor: Contribution(data=[], onDonationFailed=Function1<kotlin.coroutines.Continuation<? super kotlin.Unit>, java.lang.Object>)
```
if it is an object it will just print the name `EmptyExposureWindowDonation` for example